Services/Devices.cs

```diff
@@ -174,15 +210,86 @@ public IDeviceClient GetClient(Device device, IoTHubProtocol protocol)
             await this.GetRegistry().UpdateTwinAsync(deviceId, twin, "*");
         }
 
-        /// <summary> 
-        /// Get IoTHub connection string from either the user provided value or the configuration 
+        /// <summary>
+        /// Create a list of devices
         /// </summary>
-        public void SetCurrentIotHub()
+        public async Task CreateListAsync(IEnumerable<string> deviceIds)
         {
-            string connString = this.connectionStringManager.GetIotHubConnectionString();
-            this.registry = RegistryManager.CreateFromConnectionString(connString);
-            this.ioTHubHostName = IotHubConnectionStringBuilder.Create(connString).HostName;
-            this.log.Info("Selected active IoT Hub for devices", () => new { this.ioTHubHostName });
+            this.SetupHub();
+
+            var batches = this.SplitArray(deviceIds.ToList(), REGISTRY_MAX_BATCH_SIZE).ToArray();
+
+            this.log.Info("Creating devices",
+                () => new { Count = deviceIds.Count(), Batches = batches.Length, REGISTRY_MAX_BATCH_SIZE });
+
+            for (var batchNumber = 0; batchNumber < batches.Length; batchNumber++)
+            {
+                var batch = batches[batchNumber];
+
+                this.log.Info("Creating devices batch",
+                    () => new { batchNumber, batchSize = batch.Count() });
+
+                BulkRegistryOperationResult result = await this.registry.AddDevices2Async(
```

Comment:
What happens if creation fails? Do we stop the simulation?

Reply:
What happens if the hub isn't available or there is something wrong w/t he iterator of devices you pass?  Does this api throw or will it always succeed and return result?

Reply:
exceptions will occur in case of timeout, wrong conn string, etc. The exceptions will bubble up to the caller, in this case `SimulationRunner.Start`, which is designed to handle them with a retry:

```csharp
try
{
    var devices = this.simulations.GetDeviceIds(simulation);

    // This will ignore existing devices, creating only the missing ones
    this.devices.CreateListAsync(devices).Wait(TimeSpan.FromSeconds(30));
}
catch (Exception e)
{
    this.running = false;
    this.starting = false;
    this.log.Error("Failed to create devices", () => new { e });

    // Return and retry
    return;
}
```

Reply:
If any of the devices passed already exists (which is fine), the operation will succeed, tracking a 409 (conflict) in the `result` variable.